subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/r34/BuildProgressCrossVersionSpec.groovy

```diff
@@ -0,0 +1,229 @@
+    def "generates plugin application events for multi-project build"() {
+        given:
+        settingsFile << """
+            rootProject.name = 'multi'
+            include 'a', 'b'
+        """
+
+        when:
+        def events = new ProgressEvents()
+        withConnection {
+            ProjectConnection connection ->
+                connection.newBuild()
+                    .addProgressListener(events)
+                    .run()
+        }
+
+        then:
+        events.assertIsABuild()
+
+        def configureRoot = events.operation("Configure project :")
+        configureRoot.child("Apply plugin id 'org.gradle.help-tasks'")
+
+        def configureA = events.operation("Configure project :a")
+        configureA.child("Apply plugin id 'org.gradle.help-tasks'")
+
+        def configureB = events.operation("Configure project :b")
+        configureB.child("Apply plugin id 'org.gradle.help-tasks'")
+    }
+
+    def "generates plugin application events when configuration fails"() {
+        given:
+        settingsFile << """
+            rootProject.name = 'multi'
+            include 'a', 'b'
+        """
+        file("a/build.gradle") << """
+            throw new RuntimeException("broken")
+"""
+
+        when:
+        def events = new ProgressEvents()
+        withConnection {
+            ProjectConnection connection ->
+                connection.newBuild()
+                    .addProgressListener(events)
+                    .run()
+        }
+
+        then:
+        def e = thrown(BuildException)
+        e.cause.message =~ /A problem occurred evaluating project ':a'/
+
+        events.assertIsABuild()
+
+        def configureBuild = events.operation("Configure build")
+        configureBuild.failed
+
+        def configureRoot = events.operation("Configure project :")
+        configureRoot.child("Apply plugin id 'org.gradle.help-tasks'")
+
+        events.operation("Configure project :a").failed
+    }
+
+    def "generates events for plugin application where project configuration is nested"() {
+        given:
+        settingsFile << """
+            rootProject.name = 'multi'
+            include 'a', 'b'
+        """.stripIndent()
+        buildFile << """
+            allprojects { apply plugin: 'java' }
+            
+            evaluationDependsOn(':a')
+        """.stripIndent()
+        file("a/build.gradle") << """
+            evaluationDependsOn(':b')
+        """.stripIndent()
+
+        when:
+        def events = new ProgressEvents()
+        withConnection {
+            ProjectConnection connection ->
+                connection.newBuild()
+                    .addProgressListener(events)
+                    .run()
+        }
+
+        then:
+        events.assertIsABuild()
+
+        def configureRoot = events.operation("Configure project :")
+        configureRoot.child("Apply plugin id 'org.gradle.help-tasks'")
+        configureRoot.children("Apply plugin id 'org.gradle.java'").size() == 3
```

Comment:
Speaking of which, we should add some coverage for plugins applied from init and settings scripts to the `Gradle` and `Settings` objects.